pkg/reconciliation/reconciliation.go

```diff
@@ -0,0 +1,53 @@
+package reconciliation
+
+import (
+	"context"
+	"time"
+
+	"github.com/k8ssandra/k8ssandra-operator/pkg/annotations"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/result"
+	errors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type k8sResource interface {
+	client.Object
+	annotations.Annotated
+	DeepCopy()
+	DeepCopyInto(out *client.Object)
+}
+
+func ReconcileObject[G k8sResource](ctx context.Context,
+	desiredObj k8sResource,
+	currentObj *k8sResource,
+	client client.Client,
+	requeueDelay time.Duration) result.ReconcileResult {
+	objectKey := types.NamespacedName{
+		Name:      desiredObj.GetName(),
+		Namespace: desiredObj.GetNamespace(),
+	}
+	annotations.AddHashAnnotation(desiredObj)
+	err := client.Get(ctx, objectKey, *currentObj)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			if err := client.Create(ctx, desiredObj); err != nil {
+				return result.Error(err)
+			}
+			return result.RequeueSoon(requeueDelay)
+		} else {
+			return result.Error(err)
+		}
+	}
+
+	if !annotations.CompareHashAnnotations(*currentObj, desiredObj) {
+		resourceVersion := (*currentObj).GetResourceVersion()
+		desiredObj.DeepCopyInto(currentObj)
+		currentObj.SetResourceVersion(resourceVersion)
+		if err := client.Update(ctx, currentObj); err != nil {
+			return result.Error(err)
+		}
+		return result.RequeueSoon(requeueDelay)
+	}
+	return result.Done()
```

Comment:
Also, don't we need logic to delete objects here if they existed and are no longer desired?

Reply:
That's a good point for sure. I wonder how we should communicate that via the function's arguments... Pass an empty object just with name/namespace?

Reply:
After a small private discussion, I think an empty object works. It's certainly better than an explicit delete function, which would push duplicate logic out to code that is reconciling objects, having to decide if it wants a reconcile, or a special reconcile of removing an object. Handling it here somehow makes the most sense.

Reply:
Having looked into this, it looks like we can't establish whether an object is empty, because there is no way to access the `.Spec` field on the `client.Object`, or express to the type system that the types of objects we will be reconciling will always have a `.Spec` field. 

There might be some way to solve this, but I think the type system is hurting us here again. We might need to handle deletions separately which is a pity.